res/controllers/Pioneer-DDJ-200-scripts.js

```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
-        var deckNo = script.deckFromGroup(group);
-        var vDeckNo = DDJ200.vDeckNo[deckNo];
-        var vgroup = "[Channel" + vDeckNo + "]";
-        var pfl = ! engine.getValue(vgroup, "pfl");
-        engine.setValue(vgroup, "pfl", pfl);
-        if (!DDJ200.fourDeckMode) {
-            midi.sendShortMsg(status, 0x54, 0x7F * pfl);  // switch pfl LED
-        }
-    }
-};
-
-DDJ200.switchLEDs = function(vDeckNo) {
-    // set LEDs of controller deck 1 or 2 according to virtual deck
-    var d = (vDeckNo % 2) ? 0 : 1;           // d = deckNo - 1
-    var vgroup = "[Channel" + vDeckNo + "]";
-    DDJ200.switchPlayLED(d, engine.getValue(vgroup, "play"));
-    midi.sendShortMsg(0x90 + d, 0x0C, 0x7F *
-                      (engine.getValue(vgroup, "cue_point") !== -1));
-    DDJ200.switchSyncLED(d, engine.getValue(vgroup, "sync_enabled"));
-    if (!DDJ200.fourDeckMode) {
-        midi.sendShortMsg(0x90 + d, 0x54,
-            0x7F * engine.getValue(vgroup, "pfl"));
-    }
-
-    for (var i = 1; i <= 8; i++) {
-        var isButtonEnabled = engine.getValue(vgroup, "hotcue_" + i + "_enabled");
-        DDJ200.switchPadLED(d, i, isButtonEnabled);
-    }
-};
+            DDJ200.leftDeck.unshift();
+            DDJ200.rightDeck.unshift();
+        };
+        DDJ200.leftDeck.reconnectComponents();
+        DDJ200.rightDeck.reconnectComponents();
+    },
+});
+
+DDJ200.Deck = function(deckNumbers, midiChannel) {
+    components.Deck.call(this, deckNumbers);
+    const theDeck = this;
+
+    // set Pioneer CDJ cue mode for all decks
+    engine.setValue(this.currentDeck, "cue_cdj", true);
+
+    this.jogWheel = new components.JogWheelBasic({
+        wheelResolution: 128, // how many ticks per revolution the jogwheel has
+        alpha: 1/8, // alpha-filter
+        beta: 1/8/32,
+        rpm: 33 + 1/3,
+        inValueScale: function(value) {
+            return value - 64;
+        },
+        inKey: "jog",
+        jogCounter: 0,
+        inputSeek: function(_channel, _control, value, _status, group) {
+            if (DDJ200.leftDeck.shiftPressed) {
+                this.jogCounter += this.inValueScale(value);
+                if (this.jogCounter > 9) {
+                    engine.setValue("[Library]", "MoveDown", true);
+                    this.jogCounter = 0;
+                } else if (this.jogCounter < -9) {
+                    engine.setValue("[Library]", "MoveUp", true);
+                    this.jogCounter = 0;
+                };
+            } else {
+                const oldPos = engine.getValue(group, "playposition");
+                // Since ‘playposition’ is normalized to unity, we need to scale by
+                // song duration in order for the jog wheel to cover the same amount
+                // of time given a constant turning angle.
+                const duration = engine.getValue(group, "duration");
+                const newPos = Math.max(0, oldPos + (this.inValueScale(value) * 0.2 / duration));
```

Comment:
you can seek into the pre-roll (and there are legit reasons for doing so), so I would remove the `Math.max` here.

Reply:
yeah thats fair I guess. I have an absolute position strip on my controller, so seeking back to the start is easy there...